azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt

```diff
@@ -0,0 +1,255 @@
+    private val scenario: OracleScenario<AF, QF, AGF>,
+    private val runName: String,
+) {
+
+    init {
+        if (oracleInstances.isEmpty()) {
+            throw IllegalArgumentException("Must have at least one oracle instance to verify against")
+        }
+    }
+
+    private val actionGenerators by lazy {
+        testInstance.runTask("create-action-generators", scenario) { implementation ->
+            val systemFactory = implementation.createSystemFactory()
+            scenario.actionGenerations(systemFactory.actionGeneratorFactory)
+        }.result!!
+    }
+
+
+    fun run(): Result<AF, QF, AGF> {
+        var remainingOracles = oracleInstances
+        val oracleResults = mutableListOf<OracleResult<AF, QF, AGF>>()
+        while (oracleResults.lastOrNull()?.verified != true && remainingOracles.isNotEmpty()) {
+            val oracle = remainingOracles.first()
+            remainingOracles = remainingOracles.drop(1)
+            Log.info("Attempting to verify scenario, oracle=$oracle")
+            val oracleResult = verifyWithOracle(oracle)
+            Log.info("Verification result, oracle=$oracle verified=${oracleResult.verified}")
+            oracleResults.add(oracleResult)
+        }
+        return Result(oracleResults)
+    }
+
+    private fun verifyWithOracle(oracleInstance: ImplementationInstance<AF, CF, QF, AGF>): OracleResult<AF, QF, AGF> {
+        val declarationValidTaskResult = isDeclarationValid(oracleInstance)
+        if (declarationValidTaskResult.result != true) {
+            return OracleResult(false, oracleInstance.instanceName, listOf(declarationValidTaskResult))
+        }
+        val buildActionsValidTaskResult = areBuildActionsValid(oracleInstance)
+        if (buildActionsValidTaskResult.result != true) {
+            return OracleResult(false,
+                oracleInstance.instanceName,
+                listOf(declarationValidTaskResult, buildActionsValidTaskResult))
+        }
+        // TODO - c/should we cache generate/query of test system
+        var generatedScenario: OracleScenario<AF, QF, AGF> = scenario
+        val generateTaskResults = actionGenerators.map { generators ->
+            val generateTaskResult = generateActions(testInstance, generatedScenario, generators)
+            val actionCreators = generateTaskResult.result ?: emptyList()
+            generatedScenario = GeneratedScenario(generatedScenario, actionCreators)
+            generateTaskResult
+        }
+        val queryTaskResult = query(testInstance, generatedScenario)
+        val answerCount = queryTaskResult.result?.size ?: 0
+        Log.debug("Query result, oracle=$oracleInstance answerCount=$answerCount")
+        if (answerCount == 0) {
+            return OracleResult(false,
+                oracleInstance.instanceName,
+                listOf(declarationValidTaskResult, buildActionsValidTaskResult) + generateTaskResults + queryTaskResult
```

Comment:
Why add `generateTaskResults` and `queryTaskResult` separately?
```suggestion
                listOf(declarationValidTaskResult, buildActionsValidTaskResult, generateTaskResults, queryTaskResult)
```

Reply:
maybe that's why you suggest a fold; your suggestion is not permitted `generateTaskResults` is plural